Refresh GPS Information Panel widget - Timeout connection error!

bettellam opened the pull request:
My GNSS receiver is having trouble connecting with the new GPS Tools. The TIMEOUT connection error blocked the connection to the GNSS receiver and I had to restart QGis to make a new connection attempt.
This PR refreshes the GPS Information Panel, cancels any values present and closes any active GNSS connection!
Further minor updates.

Hi @nyalldawson
I would like to send you my feedback for the new GPS Tools!
I state immediately that not knowing the specifics of the GPS Tools lender, you can safely consider my observations null and void:

1) the Z coordinate of points does NOT use the height of the ellipsoid but uses the height (geoid) which is an approximate NMEA value that does not take into account the Undulation Model of Local Geoid.
**Conversions to Geoid Height always start from Ellipsoid Height!**
I think it's more correct to use Ellipsoidal Height for Z coordinate when saving a single point to current layer or storing a navigation track! (of course I would do it in LOG file).

2) function: LOG TO GEOPACKAGE/SPATIALITE... : Layer: GPS_POINT
a) I get double/triple points with the same timestamp even though my GNSS receiver is set to 1Hz (1 point per second)!!!!
b) I would add to the LOG: the QUALITY field: it is a very important parameter in a topographic survey.
c) It would be nice (not essential) to have the number of satellites in use for each constellation in the table.

3) I would add in Setting (and in LOG table): POLE_HEIGHT field: generally in surveys the GNSS receiver is mounted on a pole.
Consequently the Z values must take into account the new parameter!

One thing I would definitely do is modernize the names of the instruments, abandoning the old wording GPS and replacing it with GNSS.

If my remarks conflict with the GPS Tools lender's specifications,
consider void what I wrote above!

Always grateful for your amazing work!!
See you soon!
Thank you

bettellam committed:
Update qgsappgpsconnection.cpp

bettellam committed:
Update qgsgpsinformationwidget.cpp

bettellam committed:
Update qgsgpsinformationwidget.cpp

bettellam committed:
Update qgsnmeaconnection.cpp

bettellam committed:
Update qgsgpsinformationwidget.cpp

bettellam committed:
Update qgsgpsinformationwidget.cpp

bettellam committed:
Update qgsgpsinformationwidget.cpp

bettellam committed:
Update qgsgpsinformationwidget.cpp

bettellam committed:
Update qgsappgpsconnection.cpp

bettellam committed:
Update qgsgpsinformationwidget.cpp

bettellam committed:
Update qgsgpsinformationwidget.cpp

bettellam committed:
Update qgsgpsinformationwidget.cpp

bettellam committed:
Update qgsgpsinformationwidget.cpp

nyalldawson commented:
@bettellam thanks for the improvements!

There's a few unrelated changes happening here. Would you mind submitting each as an individual PR so that they can be reviewed seperately?


> My GNSS receiver is having trouble connecting with the new GPS Tools. The TIMEOUT connection error blocked the connection to the GNSS receiver and I had to restart QGis to make a new connection attempt.
This PR refreshes the GPS Information Panel, cancels any values present and closes any active GNSS connection!
Further minor updates.

This sounds reasonable, but I need to investigate whether removing the connection will cause ownership issues. I'll do that in the separate PR when you file it.

> the Z coordinate of points does NOT use the height of the ellipsoid but uses the height (geoid) which is an approximate NMEA value that does not take into account the Undulation Model of Local Geoid.
Conversions to Geoid Height always start from Ellipsoid Height!
I think it's more correct to use Ellipsoidal Height for Z coordinate when saving a single point to current layer or storing a navigation track! (of course I would do it in LOG file).

I think we should expose this as an option in the GPS settings dialog. I'd be concerned with just changing over to use ellipsoid height always as it may affect existing user workflows.

> function: LOG TO GEOPACKAGE/SPATIALITE... : Layer: GPS_POINT
a) I get double/triple points with the same timestamp even though my GNSS receiver is set to 1Hz (1 point per second)!!!!

You should be getting one point per received sentence. We should probably make this filter out so that if the row will be identical then we avoid the skip the second sentence.

> b) I would add to the LOG: the QUALITY field: it is a very important parameter in a topographic survey.

Sounds good!

> c) It would be nice (not essential) to have the number of satellites in use for each constellation in the table.

Also sounds good to me -- I designed the logging to be quite straightforward to add extra fields into, so feel free to give this a shot!

> I would add in Setting (and in LOG table): POLE_HEIGHT field: generally in surveys the GNSS receiver is mounted on a pole.
Consequently the Z values must take into account the new parameter!

+1 from me!

> One thing I would definitely do is modernize the names of the instruments, abandoning the old wording GPS and replacing it with GNSS.

Also +1 -- but keep in mind that we cannot change the internal QGIS API to replace GPS with GNSS - that needs to wait till QGIS 4.0. See https://github.com/qgis/qgis4.0_api/issues/124

bettellam commented:
Ciao @nyalldawson
Thank you for your answer!
PR is quite simple (even though I did 13 commits, I panicked....I couldn't find/see a code indent error)
1) qgsappgpsconnection.cpp
I added in :
`void QgsAppGpsConnection::onTimeOut()`
in code present in `void QgsAppGpsConnection::disconnectGps()`
in this way the connection status is canceled and the initial status is returned;

2) qgsgpsinformationwidget.cpp
`void QgsGpsInformationWidget::gpsConnecting()`
I have completed the cleaning of GNSS status values;
`void QgsGpsInformationWidget::gpsDisconnected()`
cleans GNSS status values;
Plot Signal cleaning;

3) qgsgpsinformationwidget.cpp
Minor upgrades : added Y-axis scale in Plot Signal.

4) qgsnmeaconnection.cpp
`void QgsNmeaConnection::processVtgSentence( const char *data, int len )`
I completed routine recovering value of: **direction**.

I'm a novice programmer, I don't have much experience with github, at my age I should only be a grandfather :-), give me some indications.
Should I close this PR and rewrite changes in 3 separate PRs?
Is it possible to reactivate Sky Plot Satellites chart, WITH_QWTPOLAR=true ?
Thank you
Ciao

PS
The priority problem to solve are double values in LOG file!
When **utcDateTime changes** the data is ready to be stored.
We can have a standard 1-10 Hz GNSS sampling frequency, there are receivers that work at double 20 Hz frequencies but I don't think they use QGis for surveys.
Always grateful for your remarkable work!!
Goodbye see you soon!
